# Prediction of customers' travel pattern

- https://machinelearningmastery.com/random-oversampling-and-undersampling-for-imbalanced-classification/

- https://towardsdatascience.com/predicting-hotel-bookings-with-user-search-parameters-8c570ab24805

- https://www.youtube.com/watch?v=U3X98xZ4_no

- https://www.youtube.com/watch?v=DQC_YE3I5ig

- https://github.com/coding-maniacs/over-under-sampling/blob/master/src/main.py

# 1)-Importing key modules

In [1]:
import warnings
warnings.filterwarnings('ignore')
# For processing
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import scipy
from collections import Counter
import datetime as dt
from datetime import datetime
import seaborn as sns
plt.rcParams["figure.figsize"] = (16, 10)
plt.rcParams["xtick.labelsize"] = 10
plt.figure(figsize=(16,10)) # this creates a figure 16 inch wide, 10 inch high
from pprint import pprint
%matplotlib inline
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [2]:
# For modeling building and tunning
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
#from sklearn.preprocessing import Imputer
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [3]:
# for evaluation

from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [4]:
from datetime import date
import datetime as dt

# 2)-Loading data

In [5]:
df = pd.read_csv('selected_feature.csv')
df.shape

(45805, 515)

In [7]:
df.head()

event_type     distance  num_family  len_jour  ts_hour  origin_ADB  \
0           0  5834.154716           7       6.0       11           0   
1           1  6525.926149           4      21.0       20           0   
2           1   469.781624           2       3.0       23           0   
3           1  1498.817537           1       3.0       15           0   
4           1  2921.339028           4       6.0       22           0   

   origin_ADL  origin_AER  origin_AGP  origin_AKL  ...  dest_YEG  dest_YMQ  \
0           0           0           0           0  ...         0         0   
1           0           0           0           0  ...         0         0   
2           0           0           0           0  ...         0         0   
3           0           0           0           0  ...         0         0   
4           0           0           0           0  ...         0         0   

   dest_YOW  dest_YTO  dest_YUL  dest_YVR  dest_YWG  dest_YYC  dest_YYZ  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   dest_ZRH  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 515 columns]

### 3)-Model Building

### 3.1)- Separate X and y

In [8]:
target=df["event_type"]
features=df.drop(['event_type'], axis=1)

In [9]:
print(target.shape)
print(features.shape)

(45805,)
(45805, 514)


### 3.2)- Normalize data

This will take care of those spike values in variables like distance

In [10]:
from sklearn.preprocessing import StandardScaler
X = StandardScaler().fit_transform(features)

### 3.3)-train_test_split

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, target, test_size=0.3, random_state=0)

In [12]:
print(X_train.shape)
print(X_test.shape)

print(y_train.shape)
print(y_test.shape)

(32063, 514)
(13742, 514)
(32063,)
(13742,)


### 3.4)-Applying Logistic Classifier

In [13]:
logreg = LogisticRegression(C=1e5)
logreg.fit(X_train, y_train)
predictions_LR = logreg.predict(X_test)

In [14]:
print(accuracy_score(y_test,predictions_LR))

0.9598311745015282


In [15]:
print(classification_report(y_test,predictions_LR))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98     13217
           1       0.03      0.00      0.00       525

    accuracy                           0.96     13742
   macro avg       0.50      0.50      0.49     13742
weighted avg       0.93      0.96      0.94     13742



Accuracy score is good but, is it enough? Looking at recall, precision does not seem to be case where search class is doing very well but, booking event type class is poor. This is where we say that accuracy isn't a good matrics to evaluate our model.

There are two problems here.

- 1- Imbalnced class 
- 2- Selecting correct evaluation matric

In [17]:
#let's review event_type again

df.event_type.value_counts(normalize=True)

0    0.960528
1    0.039472
Name: event_type, dtype: float64

In [18]:
df.event_type.value_counts()

0    43997
1     1808
Name: event_type, dtype: int64

Book class has only 3.9 % of values i.e 1808. This is why our model did well for accuracy. As accuracy is overall picture how model predicts correct values. Unfortunately those correct values are search class samples. We are interested in onversion likelihood i.e search to booking cases. 

Luckily, there is solution to this 

# 4)- Solving imbalnced Class Problem

**END OF NOTEBOOK3**